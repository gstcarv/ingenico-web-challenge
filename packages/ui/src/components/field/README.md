# üìù Field Component

A flexible form field wrapper component that provides consistent layout and styling for form elements with support for labels and validation messages.

## üì¶ Installation

```bash
import { Field } from '@ingenico-challenge/ui'
```

## üöÄ Usage

### Basic Field with Label

```tsx
import { Field } from '@ingenico-challenge/ui'

function MyForm() {
  return (
    <Field>
      <Field.Label>Email Address</Field.Label>
      <input type="email" placeholder="Enter your email" />
    </Field>
  )
}
```

### Field with Validation

```tsx
import { Field } from '@ingenico-challenge/ui'

function FormWithValidation() {
  const [error, setError] = useState('')

  return (
    <Field>
      <Field.Label>Password</Field.Label>
      <input 
        type="password" 
        onChange={(e) => {
          if (e.target.value.length < 8) {
            setError('Password must be at least 8 characters')
          } else {
            setError('')
          }
        }}
      />
      {error && <Field.Validation>{error}</Field.Validation>}
    </Field>
  )
}
```

### Field with Custom Styling

```tsx
<Field className="max-w-md">
  <Field.Label className="text-blue-600">Custom Label</Field.Label>
  <input type="text" />
  <Field.Validation className="text-red-500 font-bold">
    Custom error styling
  </Field.Validation>
</Field>
```

### Field with React Hook Form

```tsx
import { useForm } from 'react-hook-form'
import { Field } from '@ingenico-challenge/ui'

function HookFormExample() {
  const { register, formState: { errors } } = useForm()

  return (
    <form>
      <Field>
        <Field.Label>Username</Field.Label>
        <input {...register('username', { required: 'Username is required' })} />
        {errors.username && (
          <Field.Validation>{errors.username.message}</Field.Validation>
        )}
      </Field>
    </form>
  )
}
```

## üìã Props

### Field Props

| Prop | Type | Default | Description |
|------|------|---------|-------------|
| `children` | `ReactNode` | - | Field content (label, input, validation) |
| `className` | `string` | - | Additional CSS classes |
| `...props` | `DivHTMLAttributes` | - | All standard div attributes |

### Field.Label Props

| Prop | Type | Default | Description |
|------|------|---------|-------------|
| `children` | `ReactNode` | - | Label text content |
| `className` | `string` | - | Additional CSS classes |
| `...props` | `LabelHTMLAttributes` | - | All standard label attributes |

### Field.Validation Props

| Prop | Type | Default | Description |
|------|------|---------|-------------|
| `children` | `ReactNode` | - | Validation message content |
| `className` | `string` | - | Additional CSS classes |
| `...props` | `SpanHTMLAttributes` | - | All standard span attributes |

## üé® Styling

The component uses Class Variance Authority (CVA) for consistent styling:

### Field Container
- Flexbox column layout
- Gap between elements (`gap-2`)
- Customizable with className

### Field.Label
- Small text size (`text-xs`)
- Neutral color (`text-neutral-500`)
- Uppercase styling
- Wide letter spacing
- Medium font weight

### Field.Validation
- Error color (`text-error-400`)
- Small text size (`text-xs`)

## ‚ôø Accessibility

- Semantic HTML structure
- Proper label-input associations
- Error message accessibility
- Screen reader friendly

## üîß Customization

### Custom Field Layout

```tsx
<Field className="grid grid-cols-2 gap-4">
  <Field.Label>First Name</Field.Label>
  <input type="text" />
</Field>
```

### Custom Label Styling

```tsx
<Field.Label className="text-lg font-bold text-blue-600">
  Custom Styled Label
</Field.Label>
```

### Custom Validation Styling

```tsx
<Field.Validation className="text-red-600 font-semibold italic">
  Custom error message styling
</Field.Validation>
```

## üß™ Testing

```tsx
import { render, screen } from '@testing-library/react'
import { Field } from './field'

test('renders field with label and input', () => {
  render(
    <Field>
      <Field.Label>Test Label</Field.Label>
      <input data-testid="test-input" />
    </Field>
  )
  
  expect(screen.getByText('Test Label')).toBeInTheDocument()
  expect(screen.getByTestId('test-input')).toBeInTheDocument()
})

test('renders validation message when present', () => {
  render(
    <Field>
      <Field.Label>Test Label</Field.Label>
      <input />
      <Field.Validation>Error message</Field.Validation>
    </Field>
  )
  
  expect(screen.getByText('Error message')).toBeInTheDocument()
})
```

## üìö Storybook

View interactive examples and documentation in Storybook:

```bash
cd packages/ui
pnpm storybook
```

Navigate to the Field component in the Storybook sidebar.

## üéØ Best Practices

1. **Always use Field.Label** for proper accessibility
2. **Place validation messages** after the input element
3. **Use semantic HTML** for form elements
4. **Provide meaningful error messages** for better UX
5. **Consider using React Hook Form** for complex form management
