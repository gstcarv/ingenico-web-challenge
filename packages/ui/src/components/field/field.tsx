import { ComponentProps, ReactNode } from "react";
import { cva } from "class-variance-authority";
import { cn } from "../../utils";

export type FieldProps = ComponentProps<"div"> & {
    children: ReactNode;
};

const fieldVariants = cva("field flex flex-col gap-2");

export const Field = ({ className, children, ...props }: FieldProps) => {
    return (
        <div className={cn(fieldVariants(), className)} {...props}>
            {children}
        </div>
    );
};

// Field.Label component
export type FieldLabelProps = ComponentProps<"label"> & {
    children: ReactNode;
};

const fieldLabelVariants = cva("text-xs text-neutral-500 uppercase tracking-wide font-medium");

Field.Label = ({ className, children, ...props }: FieldLabelProps) => {
    return (
        <label className={cn(fieldLabelVariants(), className)} {...props}>
            {children}
        </label>
    );
};
