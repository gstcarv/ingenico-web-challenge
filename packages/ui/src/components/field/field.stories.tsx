import type { Meta, StoryObj } from "@storybook/react-vite";
import { Field } from "./field";
import { Select } from "../select";

const meta: Meta<typeof Field> = {
    title: "Components/Field",
    component: Field,
    parameters: {
        layout: "centered",
    },
    tags: ["autodocs"],
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
    args: {
        children: (
            <>
                <Field.Label>From</Field.Label>
                <Select>
                    <Select.Option value="">Select currency</Select.Option>
                    <Select.Option value="USD">USD</Select.Option>
                    <Select.Option value="EUR">EUR</Select.Option>
                    <Select.Option value="GBP">GBP</Select.Option>
                </Select>
            </>
        ),
    },
};

export const WithCustomLabel: Story = {
    args: {
        children: (
            <>
                <Field.Label className="text-sm text-blue-600">Custom Label</Field.Label>
                <Select>
                    <Select.Option value="">Select currency</Select.Option>
                    <Select.Option value="USD">USD</Select.Option>
                    <Select.Option value="EUR">EUR</Select.Option>
                </Select>
            </>
        ),
    },
};

export const MultipleFields: Story = {
    args: {
        children: (
            <>
                <Field.Label>From</Field.Label>
                <Select>
                    <Select.Option value="">Select currency</Select.Option>
                    <Select.Option value="USD">USD</Select.Option>
                    <Select.Option value="EUR">EUR</Select.Option>
                </Select>
            </>
        ),
    },
    render: () => (
        <div className="space-y-4">
            <Field>
                <Field.Label>From</Field.Label>
                <Select>
                    <Select.Option value="">Select currency</Select.Option>
                    <Select.Option value="USD">USD</Select.Option>
                    <Select.Option value="EUR">EUR</Select.Option>
                </Select>
            </Field>

            <Field>
                <Field.Label>To</Field.Label>
                <Select>
                    <Select.Option value="">Select currency</Select.Option>
                    <Select.Option value="EUR">EUR</Select.Option>
                    <Select.Option value="GBP">GBP</Select.Option>
                </Select>
            </Field>
        </div>
    ),
};
