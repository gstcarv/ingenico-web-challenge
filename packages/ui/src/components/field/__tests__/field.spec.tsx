import { render, screen } from "@testing-library/react";
import { describe, expect, it } from "vitest";
import { Field } from "../field";

describe("Field", () => {
    it("renders children correctly", () => {
        render(
            <Field>
                <div data-testid="child">Test content</div>
            </Field>,
        );

        expect(screen.getByTestId("child")).toBeInTheDocument();
        expect(screen.getByTestId("child")).toHaveTextContent("Test content");
    });

    it("applies custom className", () => {
        const { container } = render(
            <Field className="custom-class">
                <div>Test content</div>
            </Field>,
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it("has correct default classes", () => {
        const { container } = render(
            <Field>
                <div>Test content</div>
            </Field>,
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it("passes additional props to div element", () => {
        render(
            <Field data-testid="field-container" aria-label="test field">
                <div>Test content</div>
            </Field>,
        );

        expect(screen.getByTestId("field-container")).toBeInTheDocument();
        expect(screen.getByLabelText("test field")).toBeInTheDocument();
    });
});

describe("Field.Label", () => {
    it("renders label text correctly", () => {
        render(<Field.Label>Test Label</Field.Label>);

        expect(screen.getByText("Test Label")).toBeInTheDocument();
    });

    it("applies custom className", () => {
        const { container } = render(<Field.Label className="custom-label-class">Test Label</Field.Label>);

        expect(container.firstChild).toMatchSnapshot();
    });

    it("has correct default classes", () => {
        const { container } = render(<Field.Label>Test Label</Field.Label>);

        expect(container.firstChild).toMatchSnapshot();
    });

    it("passes additional props to label element", () => {
        render(
            <Field.Label htmlFor="test-input" data-testid="test-label">
                Test Label
            </Field.Label>,
        );

        const label = screen.getByTestId("test-label");
        expect(label).toBeInTheDocument();
        expect(label).toHaveAttribute("for", "test-input");
    });

    it("renders as a label element", () => {
        const { container } = render(<Field.Label>Test Label</Field.Label>);

        expect(container.firstChild?.nodeName).toBe("LABEL");
    });
});

describe("Field with Field.Label integration", () => {
    it("renders Field with Field.Label correctly", () => {
        render(
            <Field>
                <Field.Label>From</Field.Label>
                <div>Input content</div>
            </Field>,
        );

        expect(screen.getByText("From")).toBeInTheDocument();
        expect(screen.getByText("Input content")).toBeInTheDocument();
    });

    it("maintains proper structure with multiple children", () => {
        const { container } = render(
            <Field>
                <Field.Label>Test Label</Field.Label>
                <div>First child</div>
                <div>Second child</div>
            </Field>,
        );

        const fieldElement = container.firstChild as HTMLElement;
        expect(fieldElement.children).toHaveLength(3);
        expect(fieldElement.children[0]).toHaveTextContent("Test Label");
        expect(fieldElement.children[1]).toHaveTextContent("First child");
        expect(fieldElement.children[2]).toHaveTextContent("Second child");
    });
});
