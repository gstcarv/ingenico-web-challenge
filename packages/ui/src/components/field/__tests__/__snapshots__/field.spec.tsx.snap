// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Field > applies custom className 1`] = `
<div
  class="field flex flex-col gap-2 custom-class"
>
  <div>
    Test content
  </div>
</div>
`;

exports[`Field > has correct default classes 1`] = `
<div
  class="field flex flex-col gap-2"
>
  <div>
    Test content
  </div>
</div>
`;

exports[`Field.Label > applies custom className 1`] = `
<label
  class="text-xs text-neutral-500 uppercase tracking-wide font-medium custom-label-class"
>
  Test Label
</label>
`;

exports[`Field.Label > has correct default classes 1`] = `
<label
  class="text-xs text-neutral-500 uppercase tracking-wide font-medium"
>
  Test Label
</label>
`;
